nullaway/src/test/java/com/uber/nullaway/NullAwayContractsBooleanTests.java

```diff
@@ -0,0 +1,436 @@
+  public void identityIsNull() {
+    helper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "class Test {",
+            "  String test(@Nullable Object o1) {",
+            "    if (Validation.identity(null == o1)) {",
+            "      // BUG: Diagnostic contains: dereferenced expression",
+            "      return o1.toString();",
+            "    } else {",
+            "      return o1.toString();",
+            "    }",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void invertNotNull() {
+    helper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "class Test {",
+            "  String test(@Nullable Object o1) {",
+            "    if (Validation.invert(null != o1)) {",
+            "      // BUG: Diagnostic contains: dereferenced expression",
+            "      return o1.toString();",
+            "    } else {",
+            "      return o1.toString();",
+            "    }",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void checkAndReturn() {
+    helper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "import org.jetbrains.annotations.Contract;",
+            "class Test {",
+            "  @Contract(\"false -> fail\")",
+            "  static boolean checkAndReturn(boolean value) {",
+            "    if (!value) {",
+            "      throw new RuntimeException();",
+            "    }",
+            "    return true;",
+            "  }",
+            "  String test1(@Nullable Object o1, @Nullable Object o2) {",
+            "    if (checkAndReturn(o1 != null) && o2 != null) {",
+            "      return o1.toString() + o2.toString();",
+            "    } else {",
+            "      return o1.toString() + ",
+            "      // BUG: Diagnostic contains: dereferenced expression",
+            "        o2.toString();",
+            "    }",
+            "  }",
+            "  boolean test2(@Nullable Object o1, @Nullable Object o2) {",
+            "    return checkAndReturn(o1 != null) && o1.toString().isEmpty();",
+            "  }",
+            "  boolean test3(@Nullable Object o1, @Nullable Object o2) {",
+            "    return checkAndReturn(o1 == null) ",
+            "      // BUG: Diagnostic contains: dereferenced expression",
+            "      && o1.toString().isEmpty();",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void contractUnreachablePath() {
+    helper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "class Test {",
+            "  String test(Object required) {",
+            "    return Validation.identity(required == null)",
```

Comment:
Can't do a full review tonight, but quick question.  Would NullAway report an error for this case before your PR?  If so I would be surprised.

Reply:
> Can't do a full review tonight

No worries, I'm not in any rush

> Would NullAway report an error for this case before your PR? If so I would be surprised.

It would not because the contract would not be parsed, and values would be derived the method parameter which is declared non-null. However, with earlier revisions of my PR, the contract handler would inspect this invocation resulting in `thenUpdates.set(accessPath, Nullness.NULLABLE);` without taking into account that `required == null` can be proven always `false`.
Commenting out this block will effectively revert the relevant piece to reproduce the issue: https://github.com/uber/NullAway/blob/5e9f10d7f9e49b8e9b4658b4c678af8bcb7995c2/nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java#L223-L227

Reply:
So, this detection is useful iff there is a case for `[x]Updates.set(accessPath, Nullness.NULLABLE);`, right? Do we have such case? I can't come up with an example where we'd want to refine an existing unchanging AP *to* `Nullness.NULLABLE`, but I could well be missing something... 🤔 

(Either way, cool work and it's a good test, we definitely don't want to introduce new errors here! But, I wonder if the idea to handle all cases and converses introduced some complexity that is not needed. We've been mostly thinking of `@Contract` callsites as a way to refine `NULLABLE` into `NONNULL`. There is an argument for allowing them to refine `NULLABLE` into `NULL` - but we generally don't track must-be-null - but overall little reason to refine to `NULLABLE`: `NULLABLE`->`NULLABLE` is _redundant_ and `NONNULL`->`NULLABLE` should be _impossible_...)

Reply:
Interestingly, test cases `invertIsNull` and `invertNotNull` rely on this to report an error, otherwise it seems the conditional is assumed to act as a null check (I haven't debugged to precisely where this occurs). However, I agree that it's better to miss a potential NPE than create noise unnecessarily.